art/attacks/inference/membership_inference.py

```diff
@@ -0,0 +1,346 @@
+            self.num_features = num_features
+        else:
+            self.num_features = num_classes
+
+        super(MembershipInferenceAttackModel, self).__init__()
+
+        self.features = nn.Sequential(
+            nn.Linear(self.num_features, 512),
+            nn.ReLU(),
+            nn.Linear(512, 100),
+            nn.ReLU(),
+            nn.Linear(100, 64),
+            nn.ReLU(),
+        )
+
+        self.labels = nn.Sequential(
+            nn.Linear(self.num_classes, 256),
+            nn.ReLU(),
+            nn.Linear(256, 64),
+            nn.ReLU(),
+        )
+
+        self.combine = nn.Sequential(
+            nn.Linear(64 * 2, 1),
+        )
+
+        self.output = nn.Sigmoid()
+
+    def forward(self, x1, l):
+        out_x1 = self.features(x1)
+        out_l = self.labels(l)
+        is_member = self.combine(torch.cat((out_x1, out_l), 1))
+        return self.output(is_member)
+
+
+class MembershipInferenceBlackBoxRuleBased(InferenceAttack):
+    """
+        Implementation of a simple, rule-based black-box membership inference attack.
+
+        This implementation uses the simple rule: if the model's prediction for a sample is correct, then it is a
+        member. Otherwise, it is not a member.
+    """
+    _estimator_requirements = [BaseEstimator]
+
+    def __init__(self, classifier: Classifier):
+        """
+        Create a MembershipInferenceBlackBoxRuleBased attack instance.
+
+        :param classifier: Target classifier.
+        """
+        super(MembershipInferenceBlackBoxRuleBased, self).__init__(estimator=classifier)
+
+    def infer(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
+        """
+        Infer membership in the training set of the target estimator.
+
+        :param x: Input records to attack.
+        :param y: True labels for `x`.
+        :return: An array holding the inferred membership status, 1 indicates a member and 0 indicates non-member.
+        """
+
+        if self.estimator.input_shape[0] != x.shape[1]:
+            raise ValueError("Shape of x does not match input_shape of classifier")
+
+        y = check_and_transform_label_format(y, len(np.unique(y)), return_one_hot=True)
+        y = np.array([np.argmax(arr) for arr in y]).reshape(-1, 1)
+        if y.shape[0] != x.shape[0]:
+            raise ValueError("Number of rows in x and y do not match")
+
+        # get model's predictions for x
+        predictions = np.array([np.argmax(arr) for arr in self.estimator.predict(x)]).reshape(-1, 1)
+        return [1 if p == y[index] else 0 for index, p in enumerate(predictions)]
+
+
+class MembershipInferenceBlackBox(InferenceAttack):
+    """
+        Implementation of a learned black-box membership inference attack.
+
+        This implementation can use as input to the learning process probabilities/logits or losses,
+        depending on the type of model and provided configuration.
+    """
+    _estimator_requirements = [BaseEstimator]
```

Comment:
Does MembershipInferenceBlackBoxRuleBased require classification models? If yes, we should add ClassifierMixin to _estimator_requirements.
```suggestion
    _estimator_requirements = [BaseEstimator, ClassifierMixin]
```